### PR TITLE
Synchronizer hipblaslt part and Equality kernel

### PR DESCRIPTION
**Pull request by Jinp800125**

### CODE REVIEW

File: clients/include/testing_matmul.hpp [hunk 1]
```diff
@@ -1343,7 +1343,7 @@ void testing_matmul(const Arguments& arg)
                         h_beta_void.push_back(&h_beta[i]);
                     }
                     for(int32_t b = 0; b < block_count; b++)
-                        CHECK_HIPBLASLT_ERROR(groupedGemmVec[b].setProblem(matmul[b],
+                        CHECK_HIPBLASLT_ERROR(groupedGemmVec[b].setProblem(handle,matmul[b],
```
Comment (KKyang) [inline clients/include/testing_matmul.hpp]: Why add a handle here?

**Review by KKyang — DISMISSED** [review]

File: clients/include/testing_matmul.hpp [hunk 2]
```diff
@@ -1533,7 +1533,7 @@ void testing_matmul(const Arguments& arg)
                 }
 
                 for(int32_t b = 0; b < block_count; b++)
-                    CHECK_HIPBLASLT_ERROR(groupedGemmVec[b].setProblem(matmul[b],
+                    CHECK_HIPBLASLT_ERROR(groupedGemmVec[b].setProblem(handle,matmul[b],
```
Comment (hcman2) [inline clients/include/testing_matmul.hpp]: should add it in a new line

File: clients/include/testing_matmul.hpp (hunk 1 above)
Comment (Jinp800125) [reply clients/include/testing_matmul.hpp]: done

File: clients/include/testing_matmul.hpp (hunk 2 above)
Comment (Jinp800125) [reply clients/include/testing_matmul.hpp]: removed

**Review by jichangjichang — CHANGES_REQUESTED** [review]
please clean up code first

File: tensilelite/Tensile/Source/lib/include/Tensile/ContractionProblem.hpp [hunk 3]
```diff
@@ -729,6 +730,15 @@ namespace Tensile
             }
         }
 
+        void setSynchronizer(DataType type, size_t length)
+        {
+            if(type != DataType::None && 1)
```
Comment (jichangjichang) [inline tensilelite/Tensile/Source/lib/include/Tensile/ContractionProblem.hpp]: @Jinp800125 why && 1?

**Comment by jichangjichang** [comment]
Could you share how to calculate the required buffer size for sync buffer here?

**Comment by Jinp800125** [comment]
> Could you share how to calculate the required buffer size for sync buffer here?
as below

WorkGroup numbers x global write batch
M/MT0 x N/MT1 x NumElementsPerThread/StoreVectorWidth

**Comment by jichangjichang** [comment]
> M/MT0 x N/MT1

The sizes you allocated for sync buffer seems is not enough for other GSU size.

File: tensilelite/Tensile/Source/lib/include/Tensile/ContractionProblem.hpp (hunk 3 above)
Comment (Jinp800125) [reply tensilelite/Tensile/Source/lib/include/Tensile/ContractionProblem.hpp]: removed

**Review by hcman2 — APPROVED** [review]

File: tensilelite/Tensile/Source/lib/include/Tensile/ContractionProblemPredicates.hpp [hunk 4]
```diff
@@ -200,6 +201,59 @@ namespace Tensile
...
+                virtual bool debugEval(ContractionProblemGemm const& problem,
+                                       std::ostream&                 stream) const override
+                {
+                    return debugEvalCmp(
+                        problem,
+                        stream,
+                        "prob",
+                        (std::ceil(static_cast<float>(problem.freeSizeA(0)) / value[0])
+                         * std::ceil(static_cast<float>(problem.freeSizeB(0)) / value[1]))
+                            * (value[2] / value[3]),
```
Comment (jichangjichang) [inline tensilelite/Tensile/Source/lib/include/Tensile/ContractionProblemPredicates.hpp]: you should sync formulat with operator()

File: tensilelite/Tensile/Source/lib/include/Tensile/ContractionProblemPredicates.hpp [hunk 5]
```diff
@@ -200,6 +201,59 @@ namespace Tensile
...
+                }
+
+                static std::string Type()
+                {
+                    return "SynchronizerSizeCheck";
+                }
+
+                virtual bool operator()(ContractionProblemGemm const& problem) const override
+                {
+                    // WorkGroup numbers x global write batch x Wave numbers
```
Comment (jichangjichang) [inline tensilelite/Tensile/Source/lib/include/Tensile/ContractionProblemPredicates.hpp]: global write batch => number of global write instruction

File: tensilelite/Tensile/Source/lib/include/Tensile/ContractionProblemPredicates.hpp (hunk 4 above)
Comment (Jinp800125) [reply tensilelite/Tensile/Source/lib/include/Tensile/ContractionProblemPredicates.hpp]: sure

File: tensilelite/Tensile/Source/lib/include/Tensile/ContractionProblemPredicates.hpp (hunk 5 above)
Comment (Jinp800125) [reply tensilelite/Tensile/Source/lib/include/Tensile/ContractionProblemPredicates.hpp]: modified

**Review by jichangjichang — APPROVED** [review]
lgtm

**Pull request merged**